pennylane/templates/subroutines/qdrift.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains template for QDrift subroutine.
+"""
+import pennylane as qml
+from pennylane.operation import Operation
+from pennylane.ops import Sum
+import numpy as np
+
+class QDrift(Operation):
+    r"""An operation representing the QDrift subroutine for the complex matrix exponential
+    of a given Hamiltonian.
+
+    The QDrift subroutine provides a method to approximate the matrix exponential of hamiltonian
+    expressed as a linear combination of terms which in general do not commute. Consider the hamiltonian
+    :math:`H = \Sigma^{N}_{j=1} h_j H_{j}`, the product formula is constructed by random sampling over the terms
+    of the Hamiltonian. With probability :math:`p_j` we will add to the product the operator
+    :math:`\exp{(\frac{i \lambda H_j}{N})}`, where :math:`\lambda = \sum_{j=1}^{N} |h_j|`.
+    On the other hand, :math:`p_j` is calculated as :math:`p_j = \frac{|h_j|}{\lambda}`.
+    The number of terms to be added to the product will be entered in the class constructor.
+
+    Args:
+        hamiltonian (Union[~.Hamiltonian, ~.Sum]): The Hamiltonian of the system.
+        time (complex): The time for which the system evolves.
+
+    Keyword Args:
+        n (int): The number of terms to be added to the product formula. Default is 1.
+        seed (int): The seed for the random number generator. Default is 42.
+    Raises:
+        TypeError: The 'hamiltonian' is not of type ~.Hamiltonian, or ~.Sum.
+        ValueError: One or more of the terms in 'hamiltonian' are not Hermitian.
+        ValueError: The number of samples 'n' is not an integer.
+
+    **Example**
+
+    .. code-block:: python3
+
+        coeffs = [0.25, 0.75]
+        ops = [qml.PauliX(0), qml.PauliZ(0)]
+        H = qml.dot(coeffs, ops)
+
+        dev = qml.device("default.qubit", wires=2)
+        @qml.qnode(dev)
+        def my_circ():
+            # Prepare some state
+            qml.Hadamard(0)
+
+            # Evolve according to H
+            qml.QDrift(H, time=1.2, n = 10)
+
+            # Measure some quantity
+            return qml.probs()
+
+    >>> my_circ()
+    [0.52825127 0.         0.47174873 0.        ]
+
+
+    .. details::
+        :title: Usage Details
+
+        We can also compute the gradient with respect to the
```

Comment:
At the moment it does not work with respect to the coefficients of the Hamiltonian. However, I don't know if it should since it is a random sampling.

Reply:
Can fixing `probs` solve the conceptual issue? i.e., by making it a `kwarg` that the user can easily compute from the Hamiltonian coefficients:

`probs = abs(coeffs) / sum(abs(coeffs))`